### ISSUE TITLE
Inconsistent decoding of filenames in Windows

### ISSUE THREAD

**Issue by gavinpc** [gittup/tup#233]
## The problem

When faced with a file whose name contains certain characters (for example,
`é`), Tup does not work as expected on Windows.  It does call the requested
shell commands with the expected filenames (i.e. the filenames that Windows
itself shows), and so generally yields the correct output, but it apparently
uses a different name for the purpose of the dependency graph and thus reports
fatal errors based on the ensuing mismatches.

## Test case

```dosbatch
The following batch file reproduces the condition:

    @ECHO OFF
    
    MD test
    CD test
    tup init
	
    rem This is not directly related to the test, but enables this batch file to
    rem actually emit a file named é.in rather than ├⌐.in, which it would do in
    rem the default codepage.  See http://stackoverflow.com/a/388500
    chcp 65001
    
    rem Is it any wonder people use Cygwin?
    ECHO ^: foreach ^*.in ^|^> TYPE %%f ^> %%o ^|^> %%B.out > Tupfile
    
    ECHO hello > "e.in"
    ECHO world > "é.in"
    
    tup
```

This creates a new Tup project with two files:

- `e.in`
- `é.in`

The project contains a single `Tupfile` with a single rule:

```
    foreach *.in |> TYPE %f > %o |> %B.out
```

Note that care is taken to avoid any "Unix emulation," i.e. cygwin commands or
the black magic inside Tup that decides to use a `sh -c` wrapper.

The script produces the following output:

    [ tup ] [0.117s] Scanning filesystem...
    [ tup ] [0.257s] Reading in new environment variables...
    [ tup ] [0.385s] Parsing Tupfiles...
     100% 1) [0.003s] .
    [ tup ] [0.389s] No files to delete.
    [ tup ] [0.390s] Generating .gitignore files...
    [ tup ] [0.568s] Executing Commands...
      50% 1) [0.018s] TYPE e.in > e.out
    * 100% 2) TYPE  .in >  .out
     *** tup errors ***
    tup error: File 'D:\home\gavin\ws\test\é.out' was written to, but is not in .tup/db. You probably should specify it as an output
     -- Delete: D:\home\gavin\ws\test\é.out
    tup error: Expected to write to file ' .out' from cmd 23 but didn't
     *** Command ID=23 ran successfully, but tup failed to save the dependencies.
     *** tup: 1 job failed.

## Findings

The key lines are these, and I copy them here as they appear in emacs:

    * 100% 2) TYPE  \351.in >  \351.out
     *** tup errors ***
    tup error: File 'D:\home\gavin\ws\test\é.out' was written to, but is not in .tup/db. You probably should specify it as an output
     -- Delete: D:\home\gavin\ws\test\é.out
    tup error: Expected to write to file '\351.out' from cmd 23 but didn't

`\351` is how emacs renders a byte that hexl-mode identifies as `e9`.  According
to [this reference](http://www.fileformat.info/info/unicode/char/e9/index.htm),
the character known as `LATIN SMALL LETTER E WITH ACUTE` has a UTF-16
representation of `x00e9`; whereas its UTF-8 equivalent is `xc3 xa9`

Since NTFS stores filenames in UTF-16, I can see where the `e9` is coming from,
(although I don't see the preceding zero byte in Tup's output).  What I don't
understand is why the filename would work in one part of the operation and not
another.

I hope this is enough to identify and reproduce the problem.  Please let me know
if there is further information I can provide.

(Originally posted at https://groups.google.com/d/topic/tup-users/PVHzhGQc7fw/discussion)

**Comment by gittup** [comment]
I think this is fixed by 7f0a6916ad6b4c37cb0394be3089ff121e23b5cb, but I wouldn't be surprised if there are still issues. I've enabled t4073-utf in Windows now, and it passes.

**Comment by gittup** [comment]
I also fixed a console issue in 31a787fa3e696951bcaf123d95a3b2ba16b653f0 where it wasn't displaying the "TYPE é.in > é.out" message in a cmd shell. I think we can close this now.
